platform/mbed_cxxsupport.h

```diff
@@ -0,0 +1,1176 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2019 ARM Limited
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#ifndef MBED_CXXSUPPORT_H
+#define MBED_CXXSUPPORT_H
+
+#include "mbed_toolchain.h"
+
+/* Mbed OS code is expected to work as C++14 or later, but
+ * we currently also want to not totally break ARM Compiler 5, which has a
+ * subset of C++11 language and no C++11 library.
+ *
+ * This adaptation file fills out missing namespace std functionality for
+ * ARM Compiler 5, and backports some post-C++14 features into namespace mbed.
+ */
+
+/* Macros that can be used to mark functions and objects that are
+ * constexpr in C++14 or later, but not in earlier versions.
+ */
+#if __cplusplus >= 201402
+#define MBED_CONSTEXPR_FN_14 constexpr
+#define MBED_CONSTEXPR_OBJ_14 constexpr
+#else
+#define MBED_CONSTEXPR_FN_14 inline
+#define MBED_CONSTEXPR_OBJ_14 const
+#endif
+
+// The template stuff in here is too confusing for astyle
+// *INDENT-OFF*
+
+/* Start with some core C++11 type trait building blocks for ARMC5 */
+#if defined __CC_ARM
+
+/* ARMC5 lacks alignof and alignas keyword - do the best we can */
+
+/* alignof */
+#define alignof(T) __alignof__(T)
+
+/* alignas(N) */
+/* Types not supported - workaround is to use alignas(alignof(T)), which is legal anyway */
+#ifndef __alignas_is_defined
+#define __alignas_is_defined
+#define alignas(N) __attribute__((aligned(N)))
+#endif
+
+namespace std {
+
+/* integral_constant */
+template <typename T, T V>
+struct integral_constant {
+    using value_type = T;
+    using type = integral_constant;
+    static constexpr T value = V;
+    constexpr operator T() const noexcept
+    {
+        return V;
+    }
+    constexpr T operator()() const noexcept
+    {
+        return V;
+    }
+};
+
+template<typename T, T V>
+constexpr T integral_constant<T, V>::value;
+
+/* true_type, false_type */
+using true_type = integral_constant<bool, true>;
+using false_type = integral_constant<bool, false>;
+
+} // namespace std
+#else
+#include <cstddef>
+#include <type_traits>
+#include <utility>
+#endif
+
+/* Add some foundational stuff from C++17 or later or TS into namespace mbed */
+namespace mbed {
```

Comment:
It works as well as could be hoped.

If the toolchain has `std::void_t` (either because it's set to C++17 or is choosing to provide it as a C++14 extension), then that `using` plops it into the global namespace. The `using namespace mbed` also plops `mbed::void_t` into the global namespace.

Everything remains fine up until the application decides to reference unqualified `void_t`. At that point, the compiler says it's ambiguous, and they have to instead select `std::void_t` or `mbed::void_t`.

That disambiguation can be achieved with a _using-declaration_: `using mbed::void_t;`.